### ISSUE TITLE
printing of roughnums considered very mildly harmful

### ISSUE THREAD

**Issue by shriram** [brownplt/pyret-lang#664]
I wonder if we're printing roughnums quite right. When we print
```
~75.92759709091287
```
that sounds like an extreme degree of precision, so an error message like
```
  Attempted to compare two Roughnums for equality, which is not allowed:
  ~75.92759709091287
  ~75.92759709091287
```
looks really odd. I wonder if we can just add a little `...` at the end:
```
  Attempted to compare two Roughnums for equality, which is not allowed:
  ~75.92759709091287...
  ~75.92759709091287...
```
meaning "there's probably stuff here, we just don't know", so it's a bit less likely that a reader will think "wait, these are obviously the same!".

This potentially has implications for input. I feel they don't have to write `...` on the input. That's just a PITA.

**Comment by ds26gte** [comment]
> This potentially has implications for input. I feel they don't have to write ... on the input.

So the `...` is not just displayed in the error message, but is part of the roughnum literal syntax? Hope we don't go there.

How about the first `...` replaced by ` ± δ₁` and the second by ` ± δ₂`. The different suffixes should be enough of a hint that the two numbers are probably unequal and only possibly equal, and that in any case they are only a tiny δ apart.

**Comment by shriram** [comment]
No, I'm explicitly saying `...` is _not_ part of the literal syntax. (Though I don't mind if a person wants to put it at the end…)

We can't use all your fancy glyphs. We're ASCII.

**Comment by schanzer** [comment]
Are we _really_ ASCII? Given that we already print 1/3 with the bar on top, it seems like that horse is already out of the barn...

**Comment by shriram** [comment]
For non-string input, yes, we are.

Also, @ds26gte, I would not assume the average reader knows what your notation means — remember, our target audience includes middle-school kids.

**Comment by schanzer** [comment]
This is utterly compelling. I'm sold on `...` !

**Comment by blerner** [comment]
There's a small part of me that's not convinced: if I have roughly one half, it'd be weird for me to see `~0.5000000000...` as it violates my implicit assumption of "show me only the digits that are meaningful".  But on the other hand, `~0.33333333333....` seems useful.  I'm not rejecting this; just mentioning an odd-ish corner case.

**Comment by schanzer** [comment]
~0.5 currently prints as `~0.5`

Why not just append `...` to the existing roughnum printout?

**Comment by blerner** [comment]
I'm totally fine with that, then.  Thanks :)

**Comment by ds26gte** [comment]
PR #1028 submitted.

**Comment by blerner** [comment]
A comment from the physics folks yesterday/today has me rethinking the premise of this issue altogether.  The scenario was asking if there was a way to "limit the number of decimal places that appear in a table column".  This request is triply ambiguous: it could mean (1) the table ought to contain the actual values but only display some of the digits; (2) the table could contain numbers edited to be truncated to some number of digits; or (3) the display should just show a limited number of digits (and so could use num-to-string-digits to truncate but abandon numbers directly).

The fact that the request was phrased so vaguely makes me wonder whether *any* _a priori_ choice of number-of-digits-to-show will ever work.  So instead I think we need to rethink how we render all numbers with fractional parts.  Maybe something like a clickable `...` to expand to the next N digits; maybe tables need something similar to do it for a column all at once; but crucially we also need a way to *collapse* digits again.  I think the `~` should remain the sole delimiter of roughness vs exactitude, and I don't think this rendering issue should be a part of our concrete input syntax at all.

**Comment by jpolitz** [comment]
The thing is, I think it's nontrivial that ~ looks so much like -, especially on projectors. It's really easy to misread it; I've made this mistake and been confused many times. In my mind, that's the main reason to add more to roughnum rendering.

**Comment by blerner** [comment]
Fair point.  I think I'd rather rely on color for that, then, and pick shades/colors that are reasonably distinct.  Racket does this, e.g. with blue and dark blue.  But I am more convinced from the physics scenario that us unilaterally picking the number of digits to display, and not making that be adjustable in the output rendering, is not serving our users particularly well.

**Comment by jpolitz** [comment]
I think the physics case needs attention to the whole experience, beyond rendering a number, if it's going to be handled without the user (or library provider) writing code to express the precision.

For example, in the physics case, it's a column or multiple columns of numbers that need to be rendered differently, so the affordance shouldn't be (just) clicking individual numbers. It could be changing the rendering of _all_ numbers, or changing the rendering of particular table columns. There's a question of if that's a setting that persists across runs or not.

In addition, if the teacher wants the precision control because they don't want their students to see the full-precision numbers (or wants a particular precision), they need to have their students change it in their own UI. That, or the correct settings need to come along with the shared program.

Some of these uses are going to be best-served by those who write support code specifying precision and generating the appropriate display and/or numbers.

I think this is an important discussion to have, not least because copying and exporting table output is going to become a more and more common operation, and it's a bit ad hoc right now. This needs to be addressed for both exact and inexact numbers.

However, I also think there's a very real, concrete, less far-reaching, and mostly orthogonal issue of not confusing folks about inexactness because of display reasons. If you think there's a good enough color for that, I'd be willing to try it. I still want to know how to tell them apart at the CLI:


![image](https://user-images.githubusercontent.com/119891/30720135-1be30340-9edb-11e7-8093-376e20bf127b.png)

**Comment by shriram** [comment]
Didn't we have this conversation in a different thread? We talked about
showing the first few and last few digits of a number, with the number of
significant digits shown in the middle (so that it's easy to compare values
by magnitude instead of wondering how many digits are being hidden). We
even mocked/discussed some designs. @blerner, are you talking about the
same kind of issue? It sounds similar, but the fact that you don't
reference it makes me wonder if I'm missing something.

**Comment by shriram** [comment]
This is an email thread labeled "precise output" dated August 7, 2017. Notable is that @blerner's first response was, “This really fails to irritate me, sorry.” (-:
